### PR TITLE
avoid probable race when creating fragments

### LINKED ISSUE #1810: flock unavailable while opening fragment on 1.2
Similar to #1669 but probably not the same root cause since it's happening on 1.2.

While importing:
2019/01/02 15:57:44 pilosa.go:195: starting field import for test_bsi_field: importing values to nodes: doing import: Server error 500 Internal Server Error body:’importing existence columns: creating view: opening fragment: opening storage: flock: resource temporarily unavailable

Disabling the existence field seems to stop the problem.

This is on a 5 node cluster running imports from the pdk.

### PR DESCRIPTION
**Pull request by seebs**
Revised this based on re-reading the code and trying to test it. It turns out that the reuse of old items in the fragments map is probably harmless, but there was an actual underlying race condition that could result in trying to reopen a fragment immediately after creating it, rather than using the just-opened fragment.

## Pull request checklist

- [X] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [X] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [X] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [X] I have resolved any merge conflicts.
- [X] I have included tests that cover my changes.
- [X] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

**Comment by seebs** [comment]
> Do you think it's worth adding a test to cover this scenario like we did in https://github.com/pilosa/pilosa/pull/1675/files ?

yes, will do

**Comment by seebs** [comment]
@travisturner, I think this is probably ready now.

So when I made the (working) patch, I got stumped trying to test it -- any test which would delay suitably to manufacture the failure under the old code would deadlock under the new code, or vice versa. Then I realized I don't need an actual lock, I just need enough delay to make the scheduler likely to run both goroutines. 10ms later, here's a test which produces the expected (but unwanted) failure with the old CreateFragmentIfNotExists, but passes with the new one.

**Comment by seebs** [comment]
`staticcheck` is wiser than i am, and points out that you shouldn't call t.Fatalf() from a goroutine. Fixed using errgroup.

File: view_internal_test.go [hunk 1]
```diff
@@ -118,3 +145,15 @@ type errorBroadcaster struct {
...
 	return errors.New("intentional error")
 }
+
+// delayBroadcaster is a broadcaster which delays briefly.
+type delayBroadcaster struct {
+	nopBroadcaster
+	delay time.Duration
+}
+
+// SendSync is an implementation of Broadcaster SendSync which always returns an error.
```
Comment (travisturner) [inline view_internal_test.go]: This comment doesn't describe the actual behavior.

Comment (seebs) [reply view_internal_test.go]: "i have checked all of the comments carefully. yes, i even found the error in the comments i missed, so i fixed that too."

[narrator] They had not checked all of the comments.

**Comment by seebs** [comment]
rebased and pushed. you'll be glad to know that, while i did indeed edit the comment on errorBroadcaster to say it delays, while leaving the comment on delayBroadcaster alone, I fixed that before committing.

File: view_internal_test.go (hunk 1 above)
Comment (travisturner) [reply view_internal_test.go]: "nice narration"

[narrator] he secretly—and jealously—wished he, too, had a narrator.

**Comment by seebs** [comment]
hmm, interesting question. i wonder... say two nodes simultaneously create shards, and notify each other. do they have to wait on any processing on the remote end before they can finish/ack the message, in some way? because if they did, if they were doing shards of the same views, they could end up each having the view locked, and unable to do things to it based on the other end's message, for instance.

i'll study this more tomorrow, but i think the existence of that attempt to fix a deadlock implies something.

possibly a fix would be to add the fragment to v.fragments whether or not the broadcast succeeds; after all, if we don't, the next time we shut down, we still end up coming up with that fragment defined, and never broadcast it.

**Comment by jaffee** [comment]
I've combed through this a few times now, and I'm pretty well convinced that this particular instance won't cause a deadlock, but I do think that having something locked, and then talking to other nodes in such a way that they require locking is something to be avoided if possible.

If you'd like to discuss the previous deadlocks and how they worked, lmk

File: view_internal_test.go [hunk 2]
```diff
@@ -78,43 +78,39 @@ func TestView_DeleteFragment(t *testing.T) {
...
+	err := creates.Wait()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatalf("unexpected error: %v", err)
 	}
 }
 
-// errorBroadcaster is a broadcaster which always returns an error.
-type errorBroadcaster struct {
+// delayBroadcaster is a broadcaster which delays briefly.
```
Comment (jaffee) [inline view_internal_test.go]: total nitpick... it's not really briefly right? "delayBroadcaster is a nopBroadcaster with a configurable delay."?

Comment (seebs) [reply view_internal_test.go]: that is also true

File: view_internal_test.go [hunk 3]
```diff
@@ -78,43 +78,39 @@ func TestView_DeleteFragment(t *testing.T) {
 }
 
 // Ensure view closes fragment after failed shard broadcast.
-func TestView_CreateFragmentError(t *testing.T) {
+func TestView_CreateFragmentRace(t *testing.T) {
+	var creates errgroup.Group
 	v := mustOpenView("i", "f", "v")
 	defer v.close()
 
 	// Use a broadcaster which intentionally fails.
```
Comment (jaffee) [inline view_internal_test.go]: this comment is no longer correct

File: view_internal_test.go [hunk 4]
```diff
@@ -78,43 +78,39 @@ func TestView_DeleteFragment(t *testing.T) {
...
+type delayBroadcaster struct {
 	nopBroadcaster
+	delay time.Duration
 }
 
-// SendSync is an implementation of Broadcaster SendSync which always returns an error.
-func (errorBroadcaster) SendSync(Message) error {
-	return errors.New("intentional error")
+// SendSync is an implementation of Broadcaster SendSync which delays for a
+// specified interval before succeeding.
```
Comment (jaffee) [inline view_internal_test.go]: (without actually doing anything)

File: view.go [hunk 5]
```diff
@@ -206,45 +205,45 @@ func (v *view) recalculateCaches() {
...
+			v.logger.Printf("broadcasting create shard: %v", err)
+		}
+		close(broadcastChan)
+	}()
+
+	// We want to wait until the broadcast is complete, but what if it
+	// takes a really long time? So we time out.
+	select {
+	case <-broadcastChan:
+	case <-time.After(50 * time.Millisecond):
```
Comment (jaffee) [inline view.go]: consider logging a warning here? sendSync is taking a long time, other nodes may not know about this shard, results may not be consistent until it finishes... etc

Comment (seebs) [reply view.go]: I'd be a bit concerned about it being spammy; I don't know how often we'd expect that to happen, or how often new shards get created. I'd say maybe... Make a version that does that and do some testing? But I'm not sure what to test it on, for workloads.

Comment (jaffee) [reply view.go]: make it a logger.Debug?

**Comment by seebs** [comment]
Have fixed the two comments. I think for now I'm inclined not to report the broadcast errors, but that's a revisitable decision. I am also pondering the implications of this for a logging strategy to coalesce similar-and-repeated messages.

**Pull request merged**